### PR TITLE
MAGENTO2-392

### PR DESCRIPTION
**Pull request by aaron-martin**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [ ] Bug fix
- [X] New feature
- [ ] This change requires a documentation update

## Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I updated the CHANGELOG.md

### CODE REVIEW

File: src/Helper/Customer/Utility.php [hunk 1]
```diff
@@ -184,9 +187,40 @@ protected function setAddresses($shopgateCustomer, $magentoCustomer)
...
+
+            if (!method_exists($mageAddress, $getter)) {
+                continue;
+            }
+
+            $fieldValue = $mageAddress->$getter();
+
+            if (empty($fieldValue)) {
+                continue;
+            }
```
Comment (augsteyer) [inline src/Helper/Customer/Utility.php]: You are kind doing a bit too much here, just check if it has data, e.g
```php
if (!$mageAddress->hasData($customFieldKey)) {
 continue; 
}
$fieldValue = $this->getData($customField);
```
This way it checks for both existence of the field and if it is not null, also no snakeCase stuff needed.

**Review by augsteyer — CHANGES_REQUESTED** [review]
Also, the branch and commits should never point to a CCP ticket, you need to move the JIRA ticket to Mage2 project as soon as you know that the fix will be in that project.

**Comment by Stereck** [comment]
> Also, the branch and commits should never point to a CCP ticket, you need to move the JIRA ticket to Mage2 project as soon as you know that the fix will be in that project.

That's true as the CCP results into a general plugin feature. 
Moved ticket to the M2 project and adjusted PR title. 
Branch name is okay for now. Will be deleted in a few anyway =)

File: src/Helper/Customer/Utility.php (hunk 1 above)
Comment (aaron-martin) [reply src/Helper/Customer/Utility.php]: It doesn't appear that $magentoAddress has a hasData method or a getData method.
When i try to use them I get the error `undefined method Magento\Customer\Model\Data\Address::hasData()`

The IDE can't find the hasData or getData methods either. 

image
image

Any idea of what I am doing wrong?

Comment (augsteyer) [reply src/Helper/Customer/Utility.php]: nothing, ur code is good. Mage is just inconsistent. They ported just setData ^_^

File: src/Helper/Customer/Utility.php [hunk 2]
```diff
@@ -184,9 +187,40 @@ protected function setAddresses($shopgateCustomer, $magentoCustomer)
...
+        $customFields = [];
+        foreach (self::ADDRESS_CUSTOM_FIELD_WHITELIST as $customFieldKey) {
+            $getter = 'get' . SimpleDataObjectConverter::snakeCaseToUpperCamelCase($customFieldKey);
+
+            if (!method_exists($mageAddress, $getter)) {
+                continue;
+            }
+
+            $fieldValue = $mageAddress->$getter();
+
```
Comment (augsteyer) [inline src/Helper/Customer/Utility.php]:
```suggestion
```

File: src/Helper/Customer/Utility.php [hunk 3]
```diff
@@ -184,9 +187,40 @@ protected function setAddresses($shopgateCustomer, $magentoCustomer)
...
+     * @param Address $mageAddress
+     *
+     * @return ShopgateOrderCustomField[]
+     */
+    protected function getShopgateAddressCustomFields($mageAddress)
+    {
+        $customFields = [];
+        foreach (self::ADDRESS_CUSTOM_FIELD_WHITELIST as $customFieldKey) {
+            $getter = 'get' . SimpleDataObjectConverter::snakeCaseToUpperCamelCase($customFieldKey);
+
```
Comment (augsteyer) [inline src/Helper/Customer/Utility.php]:
```suggestion
```

**Review by augsteyer — APPROVED** [review]

File: src/Helper/Customer/Utility.php (hunk 1 above)
Comment (Stereck) [reply src/Helper/Customer/Utility.php]: Yeah, maybe they wan't to stop using magic and enforce to use class methods as they could be overwritten internally or by plugins to modify default setter/getter.

**Review by Stereck — APPROVED** [review]

**Comment by aaron-martin** [comment]
@Stereck @augsteyer 
I added similar custom field functionality when building the shopgate customer object.

File: src/Helper/Customer/Utility.php [hunk 4]
```diff
@@ -184,9 +192,41 @@ protected function setAddresses($shopgateCustomer, $magentoCustomer)
...
         $shopgateCustomer->setAddresses($addresses);
     }
+
+    /**
+     * @param CustomAttributesDataInterface $mageAddress
+     * @param string[]                      $customFieldKeys
+     *
+     * @return ShopgateOrderCustomField[]
+     */
+    protected function getShopgateCustomFields($mageAddress, $customFieldKeys)
```
Comment (Stereck) [inline src/Helper/Customer/Utility.php]: $mageAddress could be $mageCustomer too, so please refactor to something more general.
What is the first common class, both objects inherit from?

**Review by Stereck — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/) and this p
 ## [Unreleased]
 ### Added
 - support for including customer data in cart validation by exposing the getCustomer method in checkCart
+- return address custom fields in call to get customer
```
Comment (Stereck) [inline CHANGELOG.md]:
```suggestion
- export custom attributes as extra fields with customer and customer address data
```

File: src/Helper/Customer/Utility.php [hunk 6]
```diff
@@ -24,21 +24,26 @@
 
 use Magento\Customer\Api\Data\CustomerInterface;
 use Magento\Customer\Model\Data\Address;
+use Magento\Framework\Api\CustomAttributesDataInterface;
```
Comment (Stereck) [inline src/Helper/Customer/Utility.php]: Move it down right before that line to group by directory
> use Magento\Framework\Api\SimpleDataObjectConverter;

Comment (aaron-martin) [reply src/Helper/Customer/Utility.php]: done

File: src/Helper/Customer/Utility.php (hunk 4 above)
Comment (aaron-martin) [reply src/Helper/Customer/Utility.php]: changed name of $mageAddress parameter. I already found the common ancestor shared by $mageAddress and $mageCustomer. I just didn't think to change the param name.

**Review by Stereck — APPROVED** [review]

**Pull request merged**